mesonbuild/options.py

```diff
@@ -941,53 +941,29 @@ def sanitize_dir_option_value(self, prefix: str, option: OptionKey, value: T.Any
         # .as_posix() keeps the posix-like file separators Meson uses.
         return value.as_posix()
 
-    def set_value(self, key: T.Union[OptionKey, str], new_value: 'T.Any', first_invocation: bool = False) -> bool:
-        key = self.ensure_and_validate_key(key)
-        if key.name == 'prefix':
-            new_value = self.sanitize_prefix(new_value)
-        elif self.is_builtin_option(key):
-            prefix = self.get_value_for('prefix')
-            assert isinstance(prefix, str)
-            new_value = self.sanitize_dir_option_value(prefix, key, new_value)
-        if key not in self.options:
-            raise MesonException(f'Unknown options: "{key.name}" not found.')
-
-        valobj = self.options[key]
-        old_value = valobj.value
-        changed = valobj.set_value(new_value)
-
-        if valobj.readonly and changed and not first_invocation:
-            raise MesonException(f'Tried to modify read only option {str(key)!r}')
-
-        if key.name == 'prefix' and first_invocation and changed:
-            assert isinstance(old_value, str)
-            self.reset_prefixed_options(old_value, new_value)
-
-        if changed:
-            self.set_dependents(key, new_value)
-
-        return changed
-
-    def set_dependents(self, key: OptionKey, value: str) -> None:
-        if key.name != 'buildtype':
-            return
+    def _set_dependents(self, key: OptionKey, value: str) -> None:
         opt, debug = self.DEFAULT_DEPENDENTS[value]
         dkey = key.evolve(name='debug')
         optkey = key.evolve(name='optimization')
         self.options[dkey].set_value(debug)
         self.options[optkey].set_value(opt)
 
-    def set_option(self, key: OptionKey, new_value: str, first_invocation: bool = False) -> bool:
-        assert isinstance(key, OptionKey)
-        # FIXME, dupe of set_value
-        # Remove one of the two before merging to master.
+    def set_option(self, key: OptionKey, new_value: ElementaryOptionValues, first_invocation: bool = False) -> bool:
         if key.name == 'prefix':
+            assert isinstance(new_value, str), 'for mypy'
             new_value = self.sanitize_prefix(new_value)
         elif self.is_builtin_option(key):
             prefix = self.get_value_for('prefix')
-            assert isinstance(prefix, str)
+            assert isinstance(prefix, str), 'for mypy'
             new_value = self.sanitize_dir_option_value(prefix, key, new_value)
-        opt = self.get_value_object_for(key)
+
+        try:
+            opt = self.get_value_object_for(key)
+        except KeyError:
+            opt = None
+        if opt is None:
+            raise MesonException(f'Unknown options: "{key!s}" not found.')
```

Comment:
It seems to me that get_value_object_for can never return None? Any time it finds one, it converts to KeyError...

Reply:
I can look again, but I think that I found in practice it does, even though I agree, as written it shouldn't. Mypy would prove this for us if we had strict null :/

Reply:
Whatever issue I was running into before where opt was `None` I can't reproduce now. I've pulled the `MesonException` into the `except` clause instead.